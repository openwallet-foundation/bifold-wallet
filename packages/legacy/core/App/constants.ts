import { Tours } from './contexts/tour/tour-context'
import { PINValidationRules } from './types/security'
import { homeTourSteps } from './components/tour/HomeTourSteps'
import { credentialsTourSteps } from './components/tour/CredentialsTourSteps'
import { credentialOfferTourSteps } from './components/tour/CredentialOfferTourSteps'
import { proofRequestTourSteps } from './components/tour/ProofRequestTourSteps'
import { AttemptLockoutConfig } from './types/attempt-lockout-config'

const lengthOfHiddenAttributes = 10
const unicodeForBulletCharacter = '\u2022'

export const dateIntFormat = 'YYYYMMDD'

export const hiddenFieldValue = Array(lengthOfHiddenAttributes).fill(unicodeForBulletCharacter).join('')
// Used to property prefix TestIDs so they can be looked up
// by on-device automated testing systems like SauceLabs.
export const testIdPrefix = 'com.ariesbifold:id/'

export enum LocalStorageKeys {
  Onboarding = 'OnboardingState',
  // FIXME: Once hooks are updated this should no longer be necessary
  RevokedCredentials = 'RevokedCredentials',
  RevokedCredentialsMessageDismissed = 'RevokedCredentialsMessageDismissed',
  Preferences = 'PreferencesState',
  Migration = 'MigrationState',
  Tours = 'ToursState',
  HistorySettingsOption = 'historySettingsOption',
  Language = 'language',
}

export enum KeychainServices {
  Salt = 'secret.wallet.salt',
  Key = 'secret.wallet.key',
  LoginAttempt = 'wallet.loginAttempt',
}

export enum EventTypes {
  ERROR_ADDED = 'ErrorAdded',
  ERROR_REMOVED = 'ErrorRemoved',
  BIOMETRY_UPDATE = 'BiometryUpdate',
  BIOMETRY_ERROR = 'BiometryError',
}

export const second = 1000
export const minute = 60000
export const hour = 3600000

// wallet timeout of 5 minutes, 0 means the wallet never locks due to inactivity
export const walletTimeout = minute * 5

/* lockout attempt rules: The base rules apply the lockout at a specified number of incorrect attempts,
 and the threshold rules apply the lockout penalty to each attempt after the threshold that falls on the attemptIncrement.
 (In this case the threshold rule applies to every 5th incorrect login after 20).
 Keys need to be multiples of the baseRulesIncrement.
5 incorrect => 1 minute lockout
10 incorrect => 10 minute lockout
15 incorrect => 1 hour lockout
20, 25, 30, etc incorrect => 1 day lockout
*/
export const attemptLockoutConfig: AttemptLockoutConfig = {
  baseRules: {
    5: minute,
    10: 10 * minute,
    15: hour,
  },
  thresholdRules: {
    threshold: 20,
    increment: 5,
    thresholdPenaltyDuration: 24 * hour,
  },
}

export const defaultAutoLockTime = 5

export const tours: Tours = {
  homeTourSteps,
  credentialsTourSteps,
  credentialOfferTourSteps,
  proofRequestTourSteps,
}

export const walletId = 'walletId'

export const minPINLength = 6

export const PINRules: PINValidationRules = {
  only_numbers: true,
  min_length: 6,
  max_length: 6,
  no_repeated_numbers: 0,
  no_repetition_of_the_two_same_numbers: false,
  no_series_of_numbers: false,
  no_even_or_odd_series_of_numbers: false,
  no_cross_pattern: false,
}

export const domain = 'didcomm://invite'

export const tourMargin = 25

export const hitSlop = { top: 44, bottom: 44, left: 44, right: 44 }

export const templateBundleStorageDirectory = 'templates'

export const templateCacheDataFileName = 'index.json'

export const templateBundleIndexFileName = 'proof-templates.json'
