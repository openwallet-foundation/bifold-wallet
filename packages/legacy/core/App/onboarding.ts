import { Screens, OnboardingTask } from './types/navigators'

import { State } from './types/state'
import { Config } from './types/config'

export const isPrefaceComplete = (didSeePreface: boolean, showPreface: boolean): OnboardingTask => {
  return { name: Screens.Preface, completed: (didSeePreface && showPreface) || !showPreface }
}

export const isUpdateCheckComplete = (): OnboardingTask => {
  return { name: Screens.UpdateAvailable, completed: true }
}

export const isOnboardingTutorialComplete = (didCompleteTutorial: boolean): OnboardingTask => {
  return { name: Screens.Onboarding, completed: didCompleteTutorial }
}

export const isTermsComplete = (didAgreeToTerms: number, termsVersion: number): OnboardingTask => {
  return { name: Screens.Terms, completed: didAgreeToTerms === termsVersion }
}

export const isPINCreationComplete = (didCreatePIN: boolean): OnboardingTask => {
  return { name: Screens.CreatePIN, completed: didCreatePIN }
}

export const isBiometryComplete = (didConsiderBiometry: boolean): OnboardingTask => {
  return { name: Screens.UseBiometry, completed: didConsiderBiometry }
}

export const isPushNotificationComplete = (
  didConsiderPushNotifications: boolean,
  enablePushNotifications: any
): OnboardingTask => {
  return {
    name: Screens.UsePushNotifications,
    completed: !enablePushNotifications || (didConsiderPushNotifications && enablePushNotifications),
  }
}

export const isNameWalletComplete = (didNameWallet: boolean, enableWalletNaming: boolean): OnboardingTask => {
  return { name: Screens.NameWallet, completed: didNameWallet || (!enableWalletNaming && !didNameWallet) }
}

export const isAuthenticationComplete = (didCreatePIN: boolean, didAuthenticate: boolean): OnboardingTask => {
  return { name: Screens.EnterPIN, completed: didAuthenticate || !didCreatePIN }
}

export const isAttemptLockoutComplete = (servedPenalty: boolean | undefined): OnboardingTask => {
  return { name: Screens.AttemptLockout, completed: servedPenalty !== false }
}

export const generateOnboardingWorkflowSteps = (
  state: State,
  config: Config,
  termsVersion: number
): Array<OnboardingTask> => {
  const {
    didSeePreface,
    didCompleteTutorial,
    didAgreeToTerms,
    didCreatePIN,
    didConsiderBiometry,
    didConsiderPushNotifications,
    didNameWallet,
  } = state.onboarding
  const { servedPenalty } = state.loginAttempt
  const { didAuthenticate } = state.authentication
  const { enableWalletNaming } = state.preferences
  const { showPreface, enablePushNotifications } = config

  return [
    isPrefaceComplete(didSeePreface, showPreface ?? false),
    isUpdateCheckComplete(),
    isOnboardingTutorialComplete(didCompleteTutorial),
    isTermsComplete(Number(didAgreeToTerms), termsVersion),
    isPINCreationComplete(didCreatePIN),
    isBiometryComplete(didConsiderBiometry),
    isPushNotificationComplete(didConsiderPushNotifications, enablePushNotifications),
    isNameWalletComplete(didNameWallet, enableWalletNaming),
    isAttemptLockoutComplete(servedPenalty),
    isAuthenticationComplete(didCreatePIN, didAuthenticate),
  ]
}
