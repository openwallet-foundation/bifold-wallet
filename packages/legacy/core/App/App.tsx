import AgentProvider from '@credo-ts/react-hooks'
import * as React from 'react'
import { useEffect, useMemo } from 'react'
import { StatusBar } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import Toast from 'react-native-toast-message'

import { animatedComponents } from './animated-components'
import ErrorModal from './components/modals/ErrorModal'
import NetInfo from './components/network/NetInfo'
import toastConfig from './components/toast/ToastConfig'
import { Container, ContainerProvider } from './container-api'
import { ActivityProvider } from './contexts/activity'
import { AnimatedComponentsProvider } from './contexts/animated-components'
import { AuthProvider } from './contexts/auth'
import { NetworkProvider } from './contexts/network'
import { StoreProvider } from './contexts/store'
import { ThemeProvider } from './contexts/theme'
import { TourProvider } from './contexts/tour/tour-provider'
import { initLanguages, initStoredLanguage, translationResources } from './localization'
import RootStack from './navigators/RootStack'
import { theme } from './theme'
import { OpenIDCredentialRecordProvider } from './modules/openid/context/OpenIDCredentialRecordProvider'
import { tours } from './constants'

const App = (system: Container): React.FC => {
  initLanguages(translationResources)

  const AppComponent = () => {
    useMemo(() => {
      initStoredLanguage().then()
    }, [])

    useEffect(() => {
      // Hide the native splash / loading screen so that our
      // RN version can be displayed.
      SplashScreen.hide()
    }, [])

    return (
      <ContainerProvider value={system}>
        <StoreProvider>
          <AgentProvider agent={undefined}>
            <OpenIDCredentialRecordProvider>
              <ThemeProvider value={theme}>
                <AnimatedComponentsProvider value={animatedComponents}>
                  <AuthProvider>
                    <NetworkProvider>
                      <ActivityProvider>
                        <StatusBar
                          hidden={false}
                          barStyle="light-content"
                          backgroundColor={theme.ColorPallet.brand.primary}
                          translucent={false}
                        />
                        <NetInfo />
                        <ErrorModal />
                        <TourProvider tours={tours} overlayColor={'gray'} overlayOpacity={0.7}>
                          <RootStack />
                        </TourProvider>
                        <Toast topOffset={15} config={toastConfig} />
                      </ActivityProvider>
                    </NetworkProvider>
                  </AuthProvider>
                </AnimatedComponentsProvider>
              </ThemeProvider>
            </OpenIDCredentialRecordProvider>
          </AgentProvider>
        </StoreProvider>
      </ContainerProvider>
    )
  }

  return AppComponent
}
export default App
