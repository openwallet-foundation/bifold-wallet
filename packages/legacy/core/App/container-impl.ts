import { Agent } from '@credo-ts/core'
import { getProofRequestTemplates } from '@hyperledger/aries-bifold-verifier'
import { DefaultOCABundleResolver } from '@hyperledger/aries-oca/build/legacy'
import { StackNavigationProp } from '@react-navigation/stack'
import { createContext, useContext } from 'react'
import { DependencyContainer } from 'tsyringe'

import * as bundle from './assets/oca-bundles.json'
import Button from './components/buttons/Button'
import defaultIndyLedgers from './configs/ledgers/indy'
import { LocalStorageKeys, PINRules } from './constants'
import { TOKENS, Container, TokenMapping } from './container-api'
import { DispatchAction, ReducerAction } from './contexts/reducers/store'
import { defaultState } from './contexts/store'
import { useNotifications } from './hooks/notifications'
import { IHistoryManager } from './modules/history'
import HistoryManager from './modules/history/context/historyManager'
import OnboardingStack from './navigators/OnboardingStack'
import { DefaultScreenOptionsDictionary } from './navigators/defaultStackOptions'
import Developer from './screens/Developer'
import Onboarding from './screens/Onboarding'
import Preface from './screens/Preface'
import Splash from './screens/Splash'
import ScreenTerms, { TermsVersion } from './screens/Terms'
import { loadLoginAttempt } from './services/keychain'
import { BifoldLogger } from './services/logger'
import { AuthenticateStackParams, Screens } from './types/navigators'
import {
  Migration as MigrationState,
  Preferences as PreferencesState,
  State,
  Onboarding as StoreOnboardingState,
  Tours as ToursState,
  PersistentState,
} from './types/state'
import OnboardingPages from './screens/OnboardingPages'
import UseBiometry from './screens/UseBiometry'
import Scan from './screens/Scan'
import PINExplainer from './screens/PINExplainer'
import HomeHeaderView from './components/views/HomeHeaderView'
import HomeFooterView from './components/views/HomeFooterView'
import EmptyList from './components/misc/EmptyList'
import Record from './components/record/Record'
import NotificationListItem from './components/listItems/NotificationListItem'
import NoNewUpdates from './components/misc/NoNewUpdates'
import PINCreateHeader from './components/misc/PINCreateHeader'
import { PersistentStorage } from './services/storage'
import { Config, HistoryEventsLoggerConfig } from './types/config'
import { Locales } from './localization'
import ContactListItem from './components/listItems/ContactListItem'
import ContactCredentialListItem from './components/listItems/ContactCredentialListItem'
import { InlineErrorPosition } from './types/error'
import { DefaultScreenLayoutOptions } from './navigators/defaultLayoutOptions'

export const defaultConfig: Config = {
  PINSecurity: { rules: PINRules, displayHelper: false },
  settings: [],
  enableChat: true,
  enableTours: false,
  supportedLanguages: [Locales.en, Locales.fr, Locales.ptBr],
  showPreface: false,
  disableOnboardingSkip: false,
  disableContactsInSettings: false,
  disableMediatorCheck: false,
  internetReachabilityUrls: ['https://clients3.google.com/generate_204'],
  whereToUseWalletUrl: 'https://example.com',
  showScanHelp: true,
  showScanButton: true,
  showDetailsInfo: true,
  contactDetailsOptions: {
    showConnectedTime: true,
    enableEditContactName: true,
    enableCredentialList: false,
  },
}
export const defaultHistoryEventsLogger: HistoryEventsLoggerConfig = {
  logAttestationAccepted: true,
  logAttestationRefused: true,
  logAttestationRemoved: true,
  logInformationSent: true,
  logInformationNotSent: true,
  logConnection: true,
  logConnectionRemoved: true,
  logAttestationRevoked: true,
  logPinChanged: true,
  logToggleBiometry: true,
}

export class MainContainer implements Container {
  public static readonly TOKENS = TOKENS
  private _container: DependencyContainer
  private log?: BifoldLogger
  private storage: PersistentStorage<PersistentState>

  public constructor(container: DependencyContainer, log?: BifoldLogger) {
    this._container = container
    this.log = log
    this.storage = new PersistentStorage(log)
  }

  public get container(): DependencyContainer {
    return this._container
  }

  public init(): Container {
    this.log?.info(`Initializing Bifold container`)

    this._container.registerInstance(TOKENS.SCREEN_PREFACE, Preface)
    this._container.registerInstance(TOKENS.SCREEN_DEVELOPER, Developer)
    this._container.registerInstance(TOKENS.SCREEN_TERMS, { screen: ScreenTerms, version: TermsVersion })
    this._container.registerInstance(TOKENS.SCREEN_SPLASH, Splash)
    this._container.registerInstance(TOKENS.SCREEN_ONBOARDING_PAGES, OnboardingPages)
    this._container.registerInstance(TOKENS.COMPONENT_PIN_CREATE_HEADER, PINCreateHeader)
    this._container.registerInstance(TOKENS.SCREEN_USE_BIOMETRY, UseBiometry)
    this._container.registerInstance(TOKENS.SCREEN_SCAN, Scan)
    this._container.registerInstance(TOKENS.SCREEN_ONBOARDING_ITEM, Onboarding)
    this._container.registerInstance(TOKENS.SCREEN_ONBOARDING, Onboarding)
    this._container.registerInstance(TOKENS.SCREEN_PIN_EXPLAINER, PINExplainer)
    this._container.registerInstance(TOKENS.STACK_ONBOARDING, OnboardingStack)
    this._container.registerInstance(TOKENS.COMP_BUTTON, Button)
    this._container.registerInstance(TOKENS.GROUP_BY_REFERENT, false)
    this._container.registerInstance(TOKENS.HISTORY_ENABLED, false)
    this._container.registerInstance(TOKENS.HISTORY_EVENTS_LOGGER, defaultHistoryEventsLogger)
    this._container.registerInstance(TOKENS.CRED_HELP_ACTION_OVERRIDES, [])
    this._container.registerInstance(TOKENS.OBJECT_SCREEN_CONFIG, DefaultScreenOptionsDictionary)
    this._container.registerInstance(TOKENS.OBJECT_LAYOUT_CONFIG, DefaultScreenLayoutOptions)
    this._container.registerInstance(TOKENS.UTIL_LOGGER, new BifoldLogger())
    this._container.registerInstance(TOKENS.UTIL_OCA_RESOLVER, new DefaultOCABundleResolver(bundle))
    this._container.registerInstance(TOKENS.UTIL_LEDGERS, defaultIndyLedgers)
    this._container.registerInstance(TOKENS.UTIL_PROOF_TEMPLATE, getProofRequestTemplates)
    this._container.registerInstance(TOKENS.UTIL_ATTESTATION_MONITOR, { useValue: undefined })
    this._container.registerInstance(TOKENS.NOTIFICATIONS, {
      useNotifications,
    })
    this._container.registerInstance(TOKENS.NOTIFICATIONS_LIST_ITEM, NotificationListItem)
    this._container.registerInstance(TOKENS.CONFIG, defaultConfig)
    this._container.registerInstance(TOKENS.COMPONENT_CRED_LIST_HEADER_RIGHT, () => null)
    this._container.registerInstance(TOKENS.COMPONENT_CRED_LIST_OPTIONS, () => null)
    this._container.registerInstance(TOKENS.COMPONENT_CRED_LIST_FOOTER, () => null)
    this._container.registerInstance(TOKENS.COMPONENT_HOME_HEADER, HomeHeaderView)
    this._container.registerInstance(TOKENS.COMPONENT_HOME_NOTIFICATIONS_EMPTY_LIST, NoNewUpdates)
    this._container.registerInstance(TOKENS.COMPONENT_HOME_FOOTER, HomeFooterView)
    this._container.registerInstance(TOKENS.COMPONENT_CRED_EMPTY_LIST, EmptyList)
    this._container.registerInstance(TOKENS.COMPONENT_RECORD, Record)
    this._container.registerInstance(TOKENS.COMPONENT_CONTACT_LIST_ITEM, ContactListItem)
    this._container.registerInstance(TOKENS.COMPONENT_CONTACT_DETAILS_CRED_LIST_ITEM, ContactCredentialListItem)
    this._container.registerInstance(TOKENS.CACHE_CRED_DEFS, [])
    this._container.registerInstance(TOKENS.CACHE_SCHEMAS, [])
    this._container.registerInstance(TOKENS.INLINE_ERRORS, {
      enabled: true,
      hasErrorIcon: true,
      position: InlineErrorPosition.Above,
    })
    this._container.registerInstance(
      TOKENS.FN_ONBOARDING_DONE,
      (dispatch: React.Dispatch<ReducerAction<unknown>>, navigation: StackNavigationProp<AuthenticateStackParams>) => {
        return () => {
          dispatch({
            type: DispatchAction.DID_COMPLETE_TUTORIAL,
          })

          navigation.navigate(Screens.Terms)
        }
      }
    )
    this._container.registerInstance(TOKENS.FN_LOAD_HISTORY, (agent: Agent<any>): IHistoryManager => {
      return new HistoryManager(agent)
    })
    this._container.registerInstance(TOKENS.CUSTOM_NAV_STACK_1, false)
    this._container.registerInstance(TOKENS.LOAD_STATE, async (dispatch: React.Dispatch<ReducerAction<unknown>>) => {
      const loadState = async <Type>(key: LocalStorageKeys, updateVal: (newVal: Type) => void) => {
        const data = await this.storage.getValueForKey(key)
        if (data) {
          // @ts-expect-error Fix complicated type error
          updateVal(data)
        }
      }

      let loginAttempt = defaultState.loginAttempt
      let preferences = defaultState.preferences
      let migration = defaultState.migration
      let tours = defaultState.tours
      let onboarding = defaultState.onboarding

      await Promise.all([
        loadLoginAttempt().then((data) => {
          if (data) {
            loginAttempt = data
          }
        }),
        loadState<PreferencesState>(LocalStorageKeys.Preferences, (val) => (preferences = val)),
        loadState<MigrationState>(LocalStorageKeys.Migration, (val) => (migration = val)),
        loadState<ToursState>(LocalStorageKeys.Tours, (val) => (tours = val)),
        loadState<StoreOnboardingState>(LocalStorageKeys.Onboarding, (val) => (onboarding = val)),
      ])

      const state = {
        loginAttempt: { ...defaultState.loginAttempt, ...loginAttempt },
        preferences: { ...defaultState.preferences, ...preferences },
        migration: { ...defaultState.migration, ...migration },
        tours: { ...defaultState.tours, ...tours },
        onboarding: { ...defaultState.onboarding, ...onboarding },
      } as State

      dispatch({ type: DispatchAction.STATE_DISPATCH, payload: [state] })
    })

    return this
  }

  public resolve<K extends keyof TokenMapping>(token: K): TokenMapping[K] {
    return this._container.resolve(token)
  }
  public resolveAll<K extends keyof TokenMapping, T extends K[]>(
    tokens: [...T]
  ): { [I in keyof T]: TokenMapping[T[I]] } {
    return tokens.map((key) => this.resolve(key)!) as { [I in keyof T]: TokenMapping[T[I]] }
  }
}

export const SystemContext = createContext<Container | undefined>(undefined)

export const SystemProvider = SystemContext.Provider

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const useSystem = () => useContext(SystemContext)!
