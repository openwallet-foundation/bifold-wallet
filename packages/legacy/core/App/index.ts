/* eslint-disable import/no-cycle */
import type { OnboardingStyleSheet } from './screens/Onboarding'

import { Agent } from '@credo-ts/core'
import AgentProvider from '@credo-ts/react-hooks'

import App from './App'
import * as components from './components'
import { Button as IButton, ButtonImpl as Button, ButtonType } from './components/buttons/Button'
import IconButton, { ButtonLocation } from './components/buttons/IconButton'
import BulletPoint from './components/inputs/BulletPoint'
import CheckBoxRow from './components/inputs/CheckBoxRow'
import ContentGradient from './components/misc/ContentGradient'
import CredentialCard from './components/misc/CredentialCard'
import InfoBox, { InfoBoxType } from './components/misc/InfoBox'
import ErrorModal from './components/modals/ErrorModal'
import NetInfo from './components/network/NetInfo'
import Record from './components/record/Record'
import InfoTextBox from './components/texts/InfoTextBox'
import Link from './components/texts/Link'
import Text from './components/texts/Text'
import { ToastType } from './components/toast/BaseToast'
import toastConfig from './components/toast/ToastConfig'
import { AttachTourStep } from './components/tour/AttachTourStep'
import { credentialOfferTourSteps } from './components/tour/CredentialOfferTourSteps'
import { credentialsTourSteps } from './components/tour/CredentialsTourSteps'
import { homeTourSteps } from './components/tour/HomeTourSteps'
import { proofRequestTourSteps } from './components/tour/ProofRequestTourSteps'
import { TourBox } from './components/tour/TourBox'
import LimitedTextInput from './components/inputs/LimitedTextInput'
import HomeFooterView from './components/views/HomeFooterView'
import KeyboardView from './components/views/KeyboardView'
import NotificationListItem from './components/listItems/NotificationListItem'
import * as contexts from './contexts'
import { AutoLockTime, ActivityProvider, useActivity } from './contexts/activity'
import { AuthProvider } from './contexts/auth'
import { NetworkProvider } from './contexts/network'
import { useTour } from './contexts/tour/tour-context'
import { TourProvider } from './contexts/tour/tour-provider'
import AttemptLockout from './screens/AttemptLockout'
import Developer from './screens/Developer'
import OnboardingPages from './screens/OnboardingPages'
import Preface from './screens/Preface'
import Splash from './screens/Splash'
import Terms from './screens/Terms'
import UseBiometry from './screens/UseBiometry'
import { loadLoginAttempt } from './services/keychain'
import { BifoldLogger } from './services/logger'
import * as types from './types'
import Scan from './screens/Scan'
import Onboarding from './screens/Onboarding'
import { DefaultScreenOptionsDictionary, useDefaultStackOptions } from './navigators/defaultStackOptions'
import { PINRules, walletTimeout } from './constants'
import { CredentialListFooterProps } from './types/credential-list-footer'
import { OpenIDCredentialRecordProvider } from './modules/openid/context/OpenIDCredentialRecordProvider'
import { defaultConfig, defaultHistoryEventsLogger } from './container-impl'
import useInitializeAgent from './hooks/initialize-agent'
import { DefaultScreenLayoutOptions } from './navigators/defaultLayoutOptions'

export * from './navigators'
export * from './services/storage'
export * from './types/attestation'
export { LocalStorageKeys } from './constants'
export * from './services/storage'
export { initLanguages, initStoredLanguage, translationResources, Locales } from './localization'
export { defaultState, mergeReducers, StoreProvider, StoreContext, useStore } from './contexts/store'
export { default as Store, DispatchAction, reducer } from './contexts/reducers/store'
export { useDeepLinks } from './hooks/deep-links'
export { Assets as ImageAssets } from './theme'
export { ThemeProvider, useTheme } from './contexts/theme'
export { AnimatedComponentsProvider, useAnimatedComponents } from './contexts/animated-components'
export { ColorPallet } from './theme'
export { animatedComponents } from './animated-components'
export { theme } from './theme'
export { useAuth } from './contexts/auth'
export { useNetwork } from './contexts/network'
export { NavigationTheme } from './theme'
export { testIdWithKey, testIdForAccessabilityLabel } from './utils/testable'
export { Screens, Stacks, TabStacks } from './types/navigators'
export { createStyles } from './screens/OnboardingPages'
export { statusBarStyleForColor, StatusBarStyles } from './utils/luminance'
export { BifoldError } from './types/error'
export { EventTypes } from './constants'
export { migrateToAskar } from './utils/migration'
export { createLinkSecretIfRequired, getAgentModules } from './utils/agent'
export { removeExistingInvitationIfRequired, connectFromScanOrDeepLink, formatTime, useCredentialConnectionLabel, getConnectionName } from './utils/helpers'
export { isValidAnonCredsCredential, getCredentialIdentifiers } from './utils/credential'
export { buildFieldsFromAnonCredsCredential } from './utils/oca'

export type { AnimatedComponents } from './animated-components'
export type {
  ISVGAssets,
  IFontAttributes,
  IInputAttributes,
  IInputs,
  ITextTheme,
  IBrandColors,
  ISemanticColors,
  INotificationColors,
  IGrayscaleColors,
  IColorPallet,
  IAssets,
  ITheme,
} from './theme'
export type { PersistentState } from './types/state'
export type { BifoldAgent } from './utils/agent'
export type { TourStep, RenderProps } from './contexts/tour/tour-context'
export type { GenericFn } from './types/fn'
export type {
  AuthenticateStackParams,
  OnboardingStackParams,
  NotificationStackParams,
  ContactStackParams,
} from './types/navigators'
export type { OnboardingStyleSheet }
export type { WalletSecret } from './types/security'
export type { ReducerAction } from './contexts/reducers/store'
export type {
  State,
  Onboarding as OnboardingState,
  LoginAttempt as LoginAttemptState,
  Preferences as PreferencesState,
  Migration as MigrationState,
  Tours as ToursState,
} from './types/state'

export type { InlineMessageProps } from './components/inputs/InlineErrorText'

export { InlineErrorPosition } from './types/error'

export type { CredentialListFooterProps }
export * from './container-api'
export { MainContainer } from './container-impl'

export type { ScreenLayoutConfig } from './types/navigators'
export type { HistoryEventsLoggerConfig } from './types/config'

export {
  App,
  Agent,
  AgentProvider,
  AuthProvider,
  NetworkProvider,
  TourProvider,
  useTour,
  AttachTourStep,
  TourBox,
  defaultConfig,
  defaultHistoryEventsLogger,
  homeTourSteps,
  credentialsTourSteps,
  credentialOfferTourSteps,
  proofRequestTourSteps,
  ButtonType,
  ButtonLocation,
  CheckBoxRow,
  CredentialCard,
  ContentGradient,
  ErrorModal,
  IconButton,
  ActivityProvider,
  useActivity,
  InfoTextBox,
  InfoBox,
  InfoBoxType,
  Link,
  AutoLockTime,
  ToastType,
  toastConfig,
  NetInfo,
  OnboardingPages,
  OpenIDCredentialRecordProvider,
  NotificationListItem,
  useDefaultStackOptions,
  useInitializeAgent,
  Splash,
  Developer,
  Terms,
  Preface,
  HomeFooterView as HomeContentView,
  UseBiometry,
  AttemptLockout,
  Record,
  Scan,
  Onboarding,
  types,
  components,
  contexts,
  Text,
  loadLoginAttempt,
  Button,
  BifoldLogger,
  LimitedTextInput,
  KeyboardView,
  BulletPoint,
  PINRules,
  walletTimeout,
  DefaultScreenOptionsDictionary,
  DefaultScreenLayoutOptions,
}
export type { IButton }
