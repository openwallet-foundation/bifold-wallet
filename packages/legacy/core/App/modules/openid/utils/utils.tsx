import { getDomainFromUrl } from '@credo-ts/core'
import { Attribute, Field } from '@hyperledger/aries-oca/build/legacy'
import { W3cCredentialDisplay } from '../types'

/**
 * Converts a camelCase string to a sentence format (first letter capitalized, rest in lower case).
 * i.e. sanitizeString("helloWorld")  // returns: 'Hello world'
 */
export function sanitizeString(str: string) {
  const result = str.replace(/([a-z0-9])([A-Z])/g, '$1 $2').replaceAll('_', ' ')
  let words = result.split(' ')
  words = words.map((word, index) => {
    if (index === 0 || word.toUpperCase() === word) {
      return word.charAt(0).toUpperCase() + word.slice(1)
    }
    return word.charAt(0).toLowerCase() + word.slice(1)
  })
  return words.join(' ')
}

export function getHostNameFromUrl(url: string) {
  try {
    return getDomainFromUrl(url)
  } catch (error) {
    throw new Error(`Error getting hostname from url: ${error}`)
  }
}

export const buildFieldsFromOpenIDTemplate = (data: { [key: string]: unknown }): Array<Field> => {
  const fields = []
  for (const key of Object.keys(data)) {
    // omit id and type
    if (key === 'id' || key === 'type') continue

    let pushedVal: string | number | null = null
    if (typeof data[key] === 'string' || typeof data[key] === 'number') {
      pushedVal = data[key] as string | number | null
    }
    fields.push(new Attribute({ name: key, value: pushedVal }))
  }
  return fields
}

export function formatDate(input: string | Date): string {
  const date = input instanceof Date ? input : new Date(input)
  return date.toLocaleString('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: true,
  })
}

export const getAttributeField = (display: W3cCredentialDisplay, searchKey: string): Attribute | undefined => {
  for (const [key, value] of Object.entries(display.attributes)) {
    if (searchKey === key) {
      return new Attribute({
        name: key,
        value: value as string | number | null,
        mimeType: typeof value === 'number' ? 'text/number' : 'text/plain',
      })
    }
  }
  return undefined
}
