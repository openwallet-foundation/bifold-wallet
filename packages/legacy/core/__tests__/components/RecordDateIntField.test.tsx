import { render } from '@testing-library/react-native'
import React from 'react'

import RecordDateIntField from '../../App/components/record/RecordDateIntField'
import { hiddenFieldValue } from '../../App/constants'
import { testIdWithKey } from '../../App/utils/testable'

describe('RecordDateIntField Component', () => {
  test('Invalid dateInt render field value as is', async () => {
    const field = {
      name: 'Test',
      format: 'YYYY-MM-DD',
      type: 'DateInt',
      value: 'invalid date',
    }
    const tree = render(<RecordDateIntField field={field} shown={true} />)

    expect(tree).toMatchSnapshot()
  })
})
test('Valid dateInt render field value in format', async () => {
  const field = {
    name: 'Test',
    format: 'YY-MM-DD',
    type: 'DateInt',
    value: '20000101',
  }
  const tree = render(<RecordDateIntField field={field} shown={true} />)

  expect(tree).toMatchSnapshot()
})
test('Hidden field should render hidden value', async () => {
  const field = {
    name: 'Test',
    format: 'YY-MM-DD',
    type: 'DateInt',
    value: '20000101',
  }
  const tree = render(<RecordDateIntField field={field} shown={false} />)

  const hiddenFieldText = tree.getByTestId(testIdWithKey('AttributeValue'))

  expect(hiddenFieldText.children[0]).toEqual(hiddenFieldValue)
})
