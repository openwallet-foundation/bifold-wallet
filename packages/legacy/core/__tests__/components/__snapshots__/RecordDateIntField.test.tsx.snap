// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordDateIntField Component Invalid dateInt render field value as is 1`] = `
<Text
  maxFontSizeMultiplier={2}
  style={
    Array [
      Object {
        "color": "#FFFFFF",
        "fontSize": 18,
        "fontWeight": "normal",
      },
      Object {
        "color": "#FFFFFF",
        "fontSize": 18,
        "fontWeight": "normal",
      },
    ]
  }
  testID="com.ariesbifold:id/AttributeValue"
>
  Record.InvalidDateinvalid date
</Text>
`;

exports[`Valid dateInt render field value in format 1`] = `
<Text
  maxFontSizeMultiplier={2}
  style={
    Array [
      Object {
        "color": "#FFFFFF",
        "fontSize": 18,
        "fontWeight": "normal",
      },
      Object {
        "color": "#FFFFFF",
        "fontSize": 18,
        "fontWeight": "normal",
      },
    ]
  }
  testID="com.ariesbifold:id/AttributeValue"
>
  Jan 1, 2000
</Text>
`;
